File: 0031-next-permutation/0031-next-permutation.rb
# @param {Integer[]} nums
# @return {Void} Do not return anything, modify nums in-place instead.
def bubble_sort(start, nums)
    (0...nums.length).each do |i|
        (start...nums.length-i-1).each do |j|
            if nums[j] > nums[j+1]
                nums[j], nums[j+1] = nums[j+1], nums[j]
            end
        end
    end
end

def next_permutation(nums)
    return if nums.length == 1
    l, r = nums.length-2, nums.length-1
    if nums[l] < nums[r]
        nums[l], nums[r] = nums[r], nums[l]
        return
    else
        last = nums.length-1
        while l >= 0 && nums[l] >= nums[r]
            l -= 1
            r -= 1
        end
        if l < 0
            bubble_sort(0, nums)
            return
        end
        while nums[last] <= nums[l]
            last -= 1
        end
        nums[l], nums[last] = nums[last], nums[l]
        bubble_sort(l+1, nums)
        return
    end
end

File: 0031-next-permutation/0031-next-permutation.go
func bubbleSort(start int, nums []int)  {
  for i:=0;i<len(nums);i++ {
    for j:=start;j<len(nums)-i-1;j++ {
      if nums[j] > nums[j+1] {
        nums[j], nums[j+1] = nums[j+1], nums[j]
      }
    }
  }
}

func nextPermutation(nums []int)  {
  if len(nums) == 1 {
    return
  }
  l, r := len(nums)-2, len(nums)-1
  if nums[l] < nums[r] {
    nums[l], nums[r] = nums[r], nums[l]
    return
  } else {
    end := len(nums)-1
    for l >= 0 && nums[l] >= nums[r] {
      l--
      r--
    }
    if l < 0 {
      bubbleSort(0, nums)
      return
    }
    for nums[end] <= nums[l] {
      end--
    }
    nums[l], nums[end] = nums[end], nums[l]
    bubbleSort(l+1, nums)
    return
  }
}